keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -727,21 +948,73 @@ condition makes it easier to determine if a failed pod should be restarted):
 - DeletionByPodGC (an orphaned Pod deleted by pod GC)
+- DeletionByKubelet (Pod deleted due to graceful node shutdown, node pressure or Pod admission errors).
+
+We introduce a Pod condition type, called `ResourceExhausted`, used by Kubelet
+to indicate that a Pod failure is caused by exhausting pod's (or container's)
+resource (memory or ephemeral-storage). We set the `reason` field to indicate
+the reason to one of the values:
+- OOMKilled (Pod is terminated by OOM killer)
+- EphemeralStorageLimitExceeded (a limit on the Pod's emphemeral-storage is exceeded).
 
 The already existing `status.conditions` field in Pod will be used by kubernetes
-control plane components (kube-scheduler and kube-controller-manager) to append
-a dedicated condition when they send the delete operation.
+components to append a dedicated condition.
 
 The API call to append the condition will be issued as a pod status update call
 before the Pod delete request. This way the Job controller will already see the
 condition when handling a failed pod.
 
 During the implementation process we are going to review the places where the
 pod delete requests are issued to modify the code to also append a meaningful
-condition with dedicated `Type`, `Reason` and `Message` fields based on the
+condition with dedicated `type`, `reason` and `message` fields based on the
 invocation context.
 
-### Interim FailureTarget condition
+We list the reason constants above in order to gain the community consensus on
+informative names, but they are purely informational. It is not supported to use the `reason` or `message` fields
+when defining a pod failure policy (see `PodFailurePolicyOnPodConditionsPattern`
+in [JobSpec API](#jobspec-api)). Supporting of matching by the `reason` field
+is a possible extension of the feature which can be implemented once there is
+use case to motivate it. However, it may create a risk of breaking compatibility
+with evolving set of reasons in use, similar to the risk of
+[evolving condition types](#evolving-condition-types).
+
+### Detection and handling of exceeded limits
+
+We modify kubelet to add the `ResourceExhausted` condition if ephemeral-storage
+limits are exceeded or the Pod is terminated by Out-Of-Memory (OOM) killer.
+
+For detection of the ephemeral-storage limits being exceeded we reuse the pre-existing
+Kubelet code responsible for detection and eviction of Pods exceeding the limits (the code
+includes detection of both exceeding of the `emptyDir` size limit and pod's
+ephemeral-storage limits defined for containers). We extend the
+eviction-triggering code to add the `ResourceExhausted` condition.
+
+In case when a pod is evicted by Kubelet due to disk pressure, but
+the storage limits are not exceeded the `DisruptionTarget` condition is added,
+but not `ResourceExhausted`.
+
+We detect that a pod was terminated due to OOM killer based on the
+container's `reason` field being equal to `OOMKilled`. This value is set on
```

Comment:
just a comment: I'm not sure if container runtimes anywhere standardized on using OOMKilled as the reason for a OOM Kill, this is just a string the CRI sets. It does appear to be done on containerd [1] and crio [2] though, so I suppose we should be good, but might be worth to note that it would good to add this as a "formal" requirement from the container runtime since this feature is taking dependency on this behavior. 

[1] https://github.com/containerd/containerd/blob/main/pkg/cri/server/helpers.go#L62

[2] https://github.com/cri-o/cri-o/blob/main/server/container_status.go#L88-L89

Reply:
Added "Broken compatibility of communicating OOM kills between container runtime and kubelet" to the Risks and Mitigations section. As a mitigation for Beta we can offer documantation. For GA we are going to discuss standarization at the CRI level and adjust the implementation accordingly based on the outcomes.

Reply:
Kicked off the discussion as a message on the CNCF Tag Runtime mailing list: [Standardization of the OOM kill communication between container runtime and kubelet
](https://lists.cncf.io/g/cncf-tag-runtime/topic/standardization_of_the_oom/94093173?p=,,,20,0,0,0::recentpostdate/sticky,,,20,2,0,94093173,previd%3D1664811793498490629,nextid%3D1637210209940337491&previd=1664811793498490629&nextid=1637210209940337491)